File: Java/Searching/BinarySearch.java
package Searching;

import Sorting.MergeSort;

// For explanantion refer BinarySearch.js
public class BinarySearch {

    public boolean binarySearch(int arr[], int val) {
        int left = 0;
        int right = arr.length - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (arr[mid] == val) {
                return true;
            } else if (val < arr[mid]) {
                right--;
            } else {
                left++;
            }
        }

        return false;
    }

    public static void main(String[] args) {
        BinarySearch binarySearch = new BinarySearch();
        MergeSort mergeSort = new MergeSort();
        int arr[] = { 10, 9, 8, 7, 6, 5, 4, 3, 2 };
        System.out.println(binarySearch.binarySearch(mergeSort.mergeSort(arr), 200));
    }

}

File: JS/Searching/BinarySearch.js
const mergeSort = require("../Sorting/MergeSort");

// idea with binary search is that we will have a sorted array and find the mid index of this array.
// if our value is present at the mid index then return true, else if our value is less than the value at mid,
// set right to mid-1 else set left to left+1.
const binarySearch = (arr, val) => {
  let left = 0;
  let right = arr.length;
  // iterate until left <= right
  while (left <= right) {
    // find mid index of array
    mid = Math.floor((left + right) / 2);
    // if our value is presnt at mid index then return true.
    if (arr[mid] == val) {
      return true;
    }
    // if value is greater than value at mid then set left to mid+1
    else if (val > arr[mid]) {
      left = mid + 1;
    }
    // else set right to mid-1.
    else {
      right = mid - 1;
    }
  }
  // if we are not returned from function inside the loop then that means our value is not present.
  // So, simply return false.
  return false;
};

let arr = [5, 2, 5, 9, 3, 100, 10, 1, 2];
console.log(binarySearch(mergeSort.mergeSort(arr), 2));
